Add tags to torch.nn.Parameter

yf225 opened the pull request:
Currently, if some parameters in a module should be optimized in a different way (i.e. use `RowWiseAdagrad` for sparse parameters, `Adagrad` for dense parameters), such training config info has to be stored somewhere else, detached from the parameters. Here is a downside to this approach:
- If the same parameter is registered in multiple modules, it's hard to make sure all modules have the same training config info for this parameter.
  - If training config info is stored at module _instance_ level, we don't have logic that copies the training config info along with the parameter.
  - If training config info is stored at module _class_ level (e.g. as a class variable), these info won't persist after pickle/unpickle.
  - If training config info is stored in a separate object, then we need to make sure this object is serialized _along with_ the module, adding an extra step after the familiar `torch.save` API.

Ideally, information about a parameter should be stored _within_ the parameter. DPER3 (Facebook's internal ranking/personalization framework) `PyTorchParameterInfo` class takes this approach by subclassing `torch.nn.Parameter` and add `tags` and `optimizer` fields to tag the parameters with training config info. However, `PyTorchParameterInfo` is a DPER3-specific construct, and vanilla PyTorch modules still don't have the ability to add `tags` to parameters. This prevents DPER3 module from taking PyTorch module as submodules, limiting PyTorch module's usefulness in ranking use cases.

In this PR I propose that we add `tags` of dict type to `torch.nn.Parameter`. DPER3 can then store necessary training config info into `tags`, and thus be able to use PyTorch module as submodules.

cc. @kennyhorror @volkhin @lupesko @henryoier @alyssawangqq @(Benjamin Au)

Commit:
Add tags to torch.nn.Parameter

Commit:
fix

Commit:
fix

mruberry commented:
Hey @yf225, sounds like a cool idea!

I have some questions about the design, though:

- What happens when models saved without tags are loaded?
- If a model with tags is saved, then loaded by an.earlier version of PyTorch that doesn't understand them, what happens?
- Do these tags work with the JIT? (Should we add a JIT-specific test?)
- Should buffers also have tags?

Commit:
add test

Commit:
add test

Commit:
add test

Commit:
add test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
remove jit test

Commit:
clean up

Commit:
add test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
fix test

Commit:
debug

Commit:
fix test

Commit:
fix test

Commit:
fix parameter impl

Commit:
fix comment

Commit:
fix comment

Commit:
fix logic

yf225 commented:
@mruberry Thanks a lot for the review! Regarding the questions:

> * What happens when models saved without tags are loaded?

I think they should have their parameters' tags populated with empty dict. I updated `TestNN.test_module_backcompat` and `TestNN.test_conv_backcompat` to test for this.

> * If a model with tags is saved, then loaded by an.earlier version of PyTorch that doesn't understand them, what happens?

I tried to load a serialized model with tags on v1.4.0, and it gives the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/miniconda3/lib/python3.7/site-packages/torch/serialization.py", line 529, in load
    return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
  File "/data/miniconda3/lib/python3.7/site-packages/torch/serialization.py", line 702, in _legacy_load
    result = unpickler.load()
TypeError: _rebuild_parameter() takes 3 positional arguments but 4 were given
```
My guess is that it's acceptable to throw an error in this case, since we explicitly don't support forward compatibility (https://discuss.pytorch.org/t/cant-get-attribute-rebuild-parameter-on-module-torch-utils/32603/3).

A slightly different question would be "If a model with _no_ tags is saved, then loaded by an earlier version of PyTorch, what happens?" I added logic in `Parameter.__reduce_ex__` to make sure this use case continues to work.

> * Do these tags work with the JIT? (Should we add a JIT-specific test?)

I suspect that things stored in `tags` might not always be understood by JIT. For example, in DPER3 use cases we sometimes pass an optimizer object as a tag, and the optimizer object itself might not be scriptable. I also think this is okay, because JIT is mostly for inference while `tags` are mostly for storing training config info, and internally when we are interacting with Training Toolkit, the `tags` are extracted and put into a container _before_ the model is traced / scripted, so not being able to store the tags in the scripted module should be okay.

> * Should buffers also have tags?

I believe there could be a need in the future - for now the `tags` in parameters mostly store training config info, which buffers don't need. In the future if we need to add tags to buffers, we might need to introduce a class like `torch.nn.Buffer` to store buffer-specific information, since buffers are now just `torch.Tensor` which doesn't have good extension points to add extra attributes.

Commit:
fix logic

ezyang commented:
I haven't really digested the problem yet. But an alternative it to maintain a (weak) map of parameters to tags, which is functionally equivalent but doesn't require you to actually store the map on the parameters themselves.

yf225 commented:
> I haven't really digested the problem yet. But an alternative it to maintain a (weak) map of parameters to tags, which is functionally equivalent but doesn't require you to actually store the map on the parameters themselves.

I think a (weak) map of parameters to tags could work, but one concern is probably where this map is stored - if it's stored at module instance level, then we need to copy over the map from `module_a` to `module_b` when we do `module_b.weight = module_a.weight`; if it's stored at module class level, then when we `torch.save` a module we need to save the map along with it, otherwise after `torch.load` we won't be able to find the tags.

albanD commented:
Hey,

After discussing this offline with @mruberry I thing I have the following concerns about this:
1) The proposed solution does not integrate well with the current torch constructs. As mentioned above, it won't work with the `jit`. But as currently implemented, it won't work with simple things like `module.cuda()` either: we assume in few places in torch.nn that Parameters are stateless and that they can be unpacked/repacked to get the same thing.
2) I don't think this particular use-case does not require modifying the core pytorch. The use-case as I understand it is to be able to mark some parameters when they are created so that the optimizer-creation code will behave differently. In that case, we already allow for arbitrary attributes to be saved on a Tensor/Parameter object. So you can already do that without changes to Parameter.  The only difference is wrt serialization:
3) The general guideline is that you should not serialize the whole model. We recommend serializing the state dict (which does *not* contain Parameter, just plain Tensors) of your model to avoid any pickle issue with serialized custom classes. When you serialize the state dict, you can also add along any extra info you want and load them as you want.

What do you think?

yf225 commented:
@albanD Thanks a lot for the concerns! Yes I can see that in general we are assuming `torch.nn.Parameter` == `torch.Tensor`, and it's probably not worth breaking the assumption by adding more fields to `torch.nn.Parameter`, since we do have a workaround for our use case.

For now, I will assign arbitrary attributes like `tags` to `torch.nn.Parameter` at runtime, and update the DPER3 serialization logic to also serialize those extra attributes. This way, `torch.nn.Parameter` class structure doesn't need to be changed, and we can still store `tags` in `torch.nn.Parameter`.
